thoth/storages/graph/postgres.py

```diff
@@ -593,85 +1361,260 @@ def retrieve_unanalyzed_python_package_versions(
 
-        if solver_name is not None:
-            solver_info = self.parse_python_solver_name(solver_name)
-            query = query.filter_by(
-                os_name=solver_info["os_name"],
-                os_version=solver_info["os_version"],
-                python_version=solver_info["python_version"],
+        query = query.offset(start_offset).limit(count)
+
+        if distinct:
+            query = query.distinct()
+
+        query = query.all()
+
+        return query
+
+    def get_unparseable_python_package_versions_count_all(
+        self,
+        *,
+        package_name: str = None,
+        package_version: str = None,
+        index_url: str = None,
+        os_name: str = None,
+        os_version: str = None,
+        python_version: str = None,
+        distinct: bool = False,
+    ) -> int:
+        """Retrieve solved Python package versions number in Thoth Database."""
+        query = self._construct_solved_python_package_versions_query(
+            package_name=package_name,
+            package_version=package_version,
+            index_url=index_url,
+            os_name=os_name,
+            os_version=os_version,
+            python_version=python_version
             )
 
-        return query.join(Solved).filter_by(error_unparseable=True).all()
+        query = query.join(Solved).filter_by(error_unparseable=True)
 
-    def get_all_python_packages_count(self, without_error: bool = True) -> int:
-        """Retrieve number of Solved Python packages stored in the graph database."""
-        query = self._session.query(PythonPackageVersion)
+        if distinct:
+            query = query.distinct()
 
-        if without_error:
-            query = query.join(Solved).filter_by(error=False)
+        query = query.count()
 
-        return query.count()
+        return query
 
-    def get_error_python_packages_count(self, *, unsolvable: bool = False, unparseable: bool = False) -> int:
-        """Retrieve number of Python packages stored in the graph database with error flag."""
-        if unsolvable is True and unparseable is True:
-            raise ValueError("Cannot query for unparseable and unsolvable at the same time")
+    def _construct_error_solved_python_package_versions_query(
+        self,
+        package_name: str = None,
+        package_version: str = None,
+        index_url: str = None,
+        os_name: str = None,
+        os_version: str = None,
+        python_version: str = None
+    ) -> Query:
+        """Construct query for solved Python packages versions functions, the query is not executed."""
+        result = self.__class__._construct_python_package_versions_query(**locals())
 
-        return (
-            self._session.query(PythonPackageVersion)
-            .join(Solved)
-            .filter_by(error=True, error_unsolvable=unsolvable, error_unparseable=unparseable)
-            .count()
-        )
+        return result
+
+    def get_error_solved_python_package_versions(
```

Comment:
I wonder if we could merge this with `get_unsolvable_*` and `get_unparseable_*` methods and provide `unparseable`, `unsolvable` parameters to this method to query for the desired ones.

* `unsolvable=False, unparseable=False` - the current behavior of `get_error_solved_python_package_versions`
* `unsolvable=True, unparseable=False` - the current behavior of the unsolvable method
* `unsolvable=False, unparseable=True` - the current behavior of the unparseable method
* `unsolvable=True, unparseable=True` - raise an error here as this is not possible

Reply:
```get_solved_python_package_versions_count_all()
69
get_error_solved_python_package_versions_count_all(unsolvable=True)
65
get_error_solved_python_package_versions_count_all(unparseable=True)
0
get_error_solved_python_package_versions_count_all()
67```

Reply:
```
get_error_solved_python_package_versions_count_all(unsolvable=True, unparseable=True)
Traceback (most recent call last):
  File "query_test.py", line 667, in <module>
    result = graph.get_error_solved_python_package_versions_count_all(unsolvable=True, unparseable=True)
  File "/home/fmurdaca/work/aicoe/storages/thoth/storages/graph/postgres.py", line 1452, in get_error_solved_python_package_versions_count_all
    raise ValueError("Cannot query for unparseable and unsolvable at the same time")
ValueError: Cannot query for unparseable and unsolvable at the same time
```

Reply:
Awesome! :+1: